proto/pbservice/healthcheck.proto

```diff
@@ -0,0 +1,144 @@
+syntax = "proto3";
+
+package pbservice;
+
+option go_package = "github.com/hashicorp/consul/proto/pbservice";
+
+import "google/protobuf/duration.proto";
+import "proto/pbcommon/common.proto";
+import "proto/pbcommon/common_oss.proto";
+
+// This fake import path is replaced by the build script with a versioned path
+import "gogoproto/gogo.proto";
+
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_getters_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+// HealthCheck represents a single check on a given node
+//
+// mog annotation:
+//
+// target=github.com/hashicorp/consul/agent/structs.HealthCheck
+// output=healthcheck.gen.go
+// name=Structs
+message HealthCheck {
+    string Node = 1;
+    string CheckID = 2 [(gogoproto.casttype) = "github.com/hashicorp/consul/types.CheckID"];
+    string Name = 3;
+    string Status = 4;                // The current check status
+    string Notes = 5;                 // Additional notes with the status
+    string Output = 6;                // Holds output of script runs
+    string ServiceID = 7;             // optional associated service
+    string ServiceName = 8;           // optional service name
+    repeated string ServiceTags = 9;  // optional service tags
+    string Type = 12;                 // Check type: http/ttl/tcp/etc
+
+    HealthCheckDefinition Definition = 10 [(gogoproto.nullable) = false];
+
+    // mog: func-to=RaftIndexToStructs func-from=NewRaftIndexFromStructs
+    common.RaftIndex RaftIndex = 11 [(gogoproto.embed) = true, (gogoproto.nullable) = false];
+
+    // mog: func-to=EnterpriseMetaToStructs func-from=NewEnterpriseMetaFromStructs
+    common.EnterpriseMeta EnterpriseMeta = 13  [(gogoproto.nullable) = false];
+}
+
+message HeaderValue {
+    repeated string Value = 1;
+}
+
+// HealthCheckDefinition of a single HealthCheck.
+//
+// mog annotation:
+//
+// target=github.com/hashicorp/consul/agent/structs.HealthCheckDefinition
+// output=healthcheck.gen.go
+// name=Structs
+message HealthCheckDefinition {
+    string HTTP = 1;
+    bool TLSSkipVerify = 2;
+
+    // mog: func-to=MapHeadersToStructs func-from=NewMapHeadersFromStructs
+    map<string, HeaderValue> Header = 3 [(gogoproto.nullable) = false];
+    string Method = 4;
+    string Body = 18;
+    string TCP = 5;
+    google.protobuf.Duration Interval = 6
+    [(gogoproto.stdduration) = true, (gogoproto.nullable) = false];
+
+    // mog: func-to=uint func-from=uint32
+    uint32 OutputMaxSize = 9;
```

Comment:
There are 10 places in this PR where the protobuf type uses `uint32` or `int32`, and the `structs` type uses `int`. There are two ways to handle this, and currently we are using both (each is used 5 times).

1. Use mog to convert between them
2. Use gogoproto to cast the type to int

I'd like to standardize on a single way of handling it. My preference is to use `mog` to convert, because we know that gogoproto is not really maintained anymore, so I think we should be attempting to use as few of its feature as possible. That way if we need to remove it in the future, it should be less work.

I'm not exactly sure how gogoproto is handling the convert from `int` -> `int32`. I tried a marshal/unmarshal round trip for and int larger than `math.MaxInt32`, and somehow it comes back from the unmarshal as the same value. But the marshaled bytes should not be able to contain an int greater than MaxInt32, right? So is that a bug in gogoproto that would break other gRPC clients? Or am I missing something.

There is a check in Unmarshal that will return an `ErrIntOverflowNode`, but it is not hitting that case in my test. I haven't taken the time to really understand all the bit shifting that is going on in this marshal/unmarshal.

Casting using `int32` could be a problem on 32 bit platforms, if we were to ever store a value greater than MaxInt32, however we could use a different func-from to handle that problem.

Reply:
+1 for using mog only. We may be forced to move away from gogoproto somewhat soon anyway as you said!